platforms/core-execution/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultFileSystemAccess.java

```diff
@@ -129,77 +119,89 @@ public Optional<HashCode> readRegularFileContentHash(String location) {
             .collect(ImmutableMap.toImmutableMap(
                 FileSystemLocationSnapshot::getAbsolutePath,
                 Function.identity()
             ));
-        FileSystemLocationSnapshot knownExactSnapshot = previouslyKnownSnapshots.get(location);
-        if (knownExactSnapshot != null) {
-            return knownExactSnapshot;
+        FileSystemLocationSnapshot unfilteredSnapshot = alreadyStoredSnapshots.get(location);
+        if (unfilteredSnapshot != null) {
+            return filterSnapshot(filter, unfilteredSnapshot);
+        } else {
+            return snapshotAndReuse(location, filter, alreadyStoredSnapshots);
         }
-        return virtualFileSystem.store(location, vfsStorer -> {
+    }
+
+    /**
+     * Takes a snapshot of the given location and filters it according to the given filter, reusing previously known snapshots.
+     */
+    private Optional<FileSystemLocationSnapshot> snapshotAndReuse(String location, SnapshottingFilter filter, ImmutableMap<String, FileSystemLocationSnapshot> previouslyKnownSnapshots) {
+        return virtualFileSystem.storeWithAction(location, vfsStorer -> {
             File file = new File(location);
             FileMetadata fileMetadata = this.stat.stat(file);
+            FileSystemLocationSnapshot unfilteredSnapshot;
             switch (fileMetadata.getType()) {
                 case RegularFile:
                     HashCode hash = hasher.hash(file, fileMetadata.getLength(), fileMetadata.getLastModified());
-                    return vfsStorer.store(new RegularFileSnapshot(location, file.getName(), hash, fileMetadata));
+                    unfilteredSnapshot = vfsStorer.store(new RegularFileSnapshot(location, file.getName(), hash, fileMetadata));
+                    break;
                 case Missing:
-                    return vfsStorer.store(new MissingFileSnapshot(location, fileMetadata.getAccessType()));
+                    unfilteredSnapshot = vfsStorer.store(new MissingFileSnapshot(location, fileMetadata.getAccessType()));
+                    break;
                 case Directory:
-                    // This will only store the captured snapshot in the VFS if the filter did not match anything
-                    return directorySnapshotter.snapshot(
+                    // This will capture a filtered snapshot, and only store the captured snapshot in the VFS
+                    // if the filter did not filter out anything.
+                    return Optional.of(directorySnapshotter.snapshot(
                         location,
                         filter.isEmpty() ? null : filter.getAsDirectoryWalkerPredicate(),
                         previouslyKnownSnapshots,
-                        vfsStorer::store);
+                        vfsStorer::store));
                 default:
                     throw new UnsupportedOperationException();
             }
+            return filterSnapshot(filter, unfilteredSnapshot);
         });
     }
```

Comment:
💭 The main change in the PR is to this method. The original `snapshot()` method used to return a mixed bag; if you git a `DirectorySnapshot`, it would be filtered, but if it was a regular file or missing snapshot, it would not, and thus we had more complexity to deal with at the call site. We are now dealing with that difference here.